### PR TITLE
Add numpy array constructors

### PR DESCRIPTION
**Pull request by pitrou**
Add support for several array constructors (the zeros(), zeros_like() etc. series as well as arange() and linspace()).
Also allow casting to np.bool_ and numba.types.bool_ (as with integer and real types).
Also, complex division now raises ZeroDivisionError when necessary.

### CODE REVIEW

**Comment by pitrou** [comment]
Moving back to "in progress" as there may be small cleanups to do.

File: numba/runtime/nrt.c [hunk 1]
```diff
@@ -208,7 +208,7 @@ static
 void nrt_internal_dtor(void *ptr, void *info) {
     NRT_Debug(nrt_debug_print("nrt_internal_dtor %p, %p\n", ptr, info));
     if (info != NULL) {
-        memset(ptr, 0, (size_t)info);  /* for safety */
+        memset(ptr, 0xDE, (size_t)info);  /* for safety */
```
Comment (seibert) [inline numba/runtime/nrt.c]: Is the choice of a non-zero memset value for debugging purposes?

Comment (pitrou) [inline numba/runtime/nrt.c]: Yes. Zeroing could actually hide bugs, especially in the zeros() implementation :)

File: numba/targets/arrayobj.py [hunk 2]
```diff
@@ -1291,25 +1295,308 @@ def _empty_nd_impl(context, builder, arrtype, shapes):
...
-@implement(numpy.empty, types.Kind(types.Integer))
-@implement(numpy.empty, types.Kind(types.BaseTuple))
-@implement(numpy.empty, types.Kind(types.Integer), types.Kind(types.Function))
-@implement(numpy.empty, types.Kind(types.BaseTuple), types.Kind(types.Function))
-def numpy_empty_nd(context, builder, sig, args):
+def _zero_fill_array(context, builder, ary):
+    """
+    Zero-fill an array.  The array must be contiguous.
+    """
+    memset = "llvm.memset.p0i8.i%d" % (types.intp.bitwidth)
```
Comment (gmarkall) [inline numba/targets/arrayobj.py]: Is it worth making a `memset` function in cgutils? It's somewhere that I'd check for it if I wanted to use `memset` from somewhere else, but might miss it in here.

**Comment by gmarkall** [comment]
I don't feel strongly about either supporting keyword args for numpy constructors as opposed to just supporting positional arguments, but this looks good to me. However, if we only support positional arguments, it should be mentioned in the manual.

Would it be worth adding `np.eye` with this as well?

**Comment by pitrou** [comment]
> However, if we only support positional arguments, it should be mentioned in the manual.

That's the same for every other supported library function, but perhaps we should mention it regardless.

> Would it be worth adding np.eye with this as well?

I don't know, it seems hard to use it without any keyword arguments...

File: numba/targets/arrayobj.py (hunk 2 above)
Comment (pitrou) [inline numba/targets/arrayobj.py]: Yes, good point.

**Comment by gmarkall** [comment]
>>Would it be worth adding np.eye with this as well?
>I don't know, it seems hard to use it without any keyword arguments...

Ah, I hadn't realised it wasn't just a synonym of `np.identity`. Although I'm not really sure why it would be hard to use without kwargs - I could imagine doing `eye(2)`, `eye(2, 3)`, `eye(2, 3, 1.5)` and `eye(2, 3, 1, int)` without too much trouble, but it is definitely a bit less convenient without being able to use kwargs. Have I missed something?

**Comment by pitrou** [comment]
I thought that people would often omit the second parameter... if not, then good, I can add it :)

**Comment by seibert** [comment]
We have kwargs now, so feel free to add np.eye.  :)

**Comment by pitrou** [comment]
Okay, I've added np.eye() as well as kwarg support for zeros() and friends.

**Comment by pitrou** [comment]
(closed by mistake!)

**Comment by Juanlu001** [comment]
Is `np.array()` planned? Naturally I could create a zeros first and then fill the values, just wondering if I can still save some lines :)

**Comment by pitrou** [comment]
> Is np.array() planned?

With which kind of arguments? if it's `np.array([1,2,3,4])`, we will first have to support Python lists :-)

**Comment by Juanlu001** [comment]
True, didn't notice Python lists were not supported yet! Still, there's the possibility to use tuples too:

```
In [46]: np.array((1, 2, 3))
Out[46]: array([1, 2, 3])
```

**Pull request merged**